### PR TITLE
[ACTION] Reddit new actions

### PR DESCRIPTION
**Pull request by vellames-turing**
**Is there a specific app this action is for?**
Reddit

**Please provide a link to the relevant API docs for the specific service / operation.**
https://www.reddit.com/dev/api/

- [x] **Submit a Comment** | Submits a new comment or reply to a message.
- [x] **Submit a Post** | Submit creates a link, a post with an image or a self-post in the subreddit with the given title. If kind is "link" or "image", then url is expected to be a valid url to link to. Otherwise, text is the body of the self-post.
- [x] **List Comments in a Post** | Retrieves a list of comments in a Post
- [x] **List Subreddits by Query** | Searches subreddits by query

**From CommonJS to ESM**

- [x] (Source) New Comments by User
- [x] (Source) New Comments on a Post
- [x] (Source) New Hot Posts on a SubReddit
- [x] (Source) New Links by User
- [x] (Source) New Links on a SubReddit
- [x] common.js
- [x] app

### CODE REVIEW

**Comment by dylburger** [comment]
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling soures, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/reddit/actions/list-comments/list-comments.mjs
Comment (js07) [inline components/reddit/actions/list-comments/list-comments.mjs:84]: Could we paginate the comments in a post using the [`after` parameter](https://www.reddit.com/dev/api/#listings)? For example:
- Add a `max` prop like in [Airtable `common-list.js`](https://github.com/PipedreamHQ/pipedream/blob/master/components/airtable/actions/common-list.js#L18)
- Paginate comments by making API requests until there are no more comments or we've retrieved `max` comments
- Optionally add an `after` prop to set as the initial `after` parameter in the request

File: components/reddit/actions/list-comments/list-comments.mjs
Comment (js07) [inline components/reddit/actions/list-comments/list-comments.mjs:9]: Do you think we should name this action "List Comments in a Post" instead so its effect is clearer, and to differentiate it from potential future "List Comments <by/in X>" actions?

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (js07) [inline components/reddit/actions/list-subreddits/list-subreddits.mjs:9]: Do you think we should name this action "List Subreddits by Query" instead so its effect is clearer, and to differentiate it from potential future "List Subreddits <by/in X>" actions?

And if we do change the action `name`, also change the `key` to follow the [`app_name`_`slug-slugified-component-name` pattern](https://pipedream.com/docs/components/guidelines/#component-key-pattern).

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (js07) [inline components/reddit/actions/list-subreddits/list-subreddits.mjs:10]:
```suggestion
  description: "List subreddits based on a search criteria. [See the docs here](https://www.reddit.com/dev/api/#GET_subreddits_search)",
```

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (js07) [inline components/reddit/actions/list-subreddits/list-subreddits.mjs:86]: The `q` parameter seems to be used to search the title _and_ description of subreddits according to the [API docs](https://www.reddit.com/dev/api/#GET_subreddits_search). Using the `subReddit` prop as the value of `q` suggests only subreddit _titles_ are searched.
Do you think we could replace the `subReddit` prop in this action with a `q` prop instead?

**Review by js07 — COMMENTED** [review]
I submitted some review comments preemptively. I'll submit the rest of the comments in a separate review!

The new actions look good so far and work for me. A few of the existing Reddit sources don't work fully because the format of `subreddit`'s `async options` return value has changed.

**Comment by vellames-turing** [comment]
@js07 Thank you for the review. I will take a look at it ASAP.

File: components/reddit/reddit.app.mjs
Comment (js07) [inline components/reddit/reddit.app.mjs:370]: Could we reuse the imported `get` from `"lodash/get.js"` instead of importing `lodash` in this file?
```js
        const isNewDataAvailable = get(redditCommunities, [
```

File: components/reddit/reddit.app.mjs
Comment (js07) [inline components/reddit/reddit.app.mjs:261]: Since the function parameters for `getNewSubredditLinks` have been changed, I think we'll need to update how `getNewSubredditLinks` is called in `sources/new-comments-on-a-post.mjs` and `sources/new-links-on-a-subreddit.mjs`.

File: components/reddit/reddit.app.mjs
Comment (js07) [inline components/reddit/reddit.app.mjs:87]: Since the value for the `subreddit` prop (using structured mode) has changed, I think we'll need to update how `this.subreddit` is used in `sources/new-links-on-a-subreddit.mjs`, `sources/new-hot-posts-on-a-subreddit.mjs`, and `sources/new-comments-on-a-post.mjs`.

**Review by js07 — CHANGES_REQUESTED** [review]

File: components/reddit/actions/list-comments/list-comments.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-comments/list-comments.mjs:13]: Just a personal preference. I will follow the reddit variable name :)

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-subreddits/list-subreddits.mjs:9]: Yeah, you're right :)

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-subreddits/list-subreddits.mjs:80]: The reddit documentation does not provide good descriptions and I was not able to figure out how exactly this works. I will remove it and maybe we can add it in a future opportunity

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-subreddits/list-subreddits.mjs:74]: I had some cases that I was not able to figure out how the variable works since Reddit does not offer good descriptions. I will remove the variables that I was not able to figure out how works

File: components/reddit/reddit.app.mjs
Comment (vellames-turing) [reply components/reddit/reddit.app.mjs:370]: Oops, I forgot to remove this one :)

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-subreddits/list-subreddits.mjs]: I was not able to figure out how this variable works as well. Even looking at google or Reddit documentation. I will remove this variable as well

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-subreddits/list-subreddits.mjs]: This is the same case as others. No description found on reddit, no example on google. I did not notice any difference in the return using true or false :/
Removing it

File: components/reddit/actions/list-comments/list-comments.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-comments/list-comments.mjs]: I think it can be a little dangerous if the user puts a high number in the limit. Reddit has a very aggressive rate limit. Besides, I think we should leave the user to paginate it the way he wants it. In Reddit you can use a `before` or `after` logic. I added these two variables and the `limit` variable. Please let me know if you disagree with my new approach

File: components/reddit/actions/list-subreddits/list-subreddits.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-subreddits/list-subreddits.mjs]: I did some tests trying to search for the description and I had bad results. Besides, the description of a subreddit is normally a big paragraph. Even if we figure out a way to make the `q` work properly for descriptions. I don't think it's worth it. Please let me know if you disagree :)

**Comment by vellames-turing** [comment]
Hey @js07. The PR is ready for review again :)

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (js07) [inline components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs:98]: I should have mentioned this before, but ideally every interaction with an app-specific API should go [through the `.app` file](https://github.com/PipedreamHQ/pipedream/blob/9789bd637e61143a5d980b35cff12758fcee5513/docs/docs/components/guidelines/README.md#methods). The `.app` file would encapsulate [knowledge about the API](https://github.com/PipedreamHQ/pipedream/blob/9789bd637e61143a5d980b35cff12758fcee5513/docs/docs/components/guidelines/README.md#good-api-usage) including API URLs and how requests are made (using `axios` vs. a client library). For example, the Twitter app's [Search Mentions](https://github.com/PipedreamHQ/pipedream/blob/master/components/twitter/sources/search-mentions/search-mentions.js) component gets data from the Twitter API through a method in the `.app` file, which can be reused in multiple components.

`spotify.app.mjs` contains [methods](https://github.com/PipedreamHQ/pipedream/blob/4d961d24f78b317186d65100a37fe0acf647ed56/components/reddit/reddit.app.js#L121-L246) used by Reddit sources, which could be reused if updated to support parameters used by the new actions.

**Review by js07 — COMMENTED** [review]
The changes look good and the sources are working for me! I left a couple comments about prop descriptions, paginating post comments, and refactoring logic of API calls into the `.app` file.

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs:108]: Indeed, there is no `after`, `before`. Sorry. We will need to use the morechildren. Maybe we should open a new issue to add this action and move on with this PR? What do you think @dannyroosevelt ?

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (vellames-turing) [reply components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs:98]: I would like to maintain this way. We cant use the `$` variable inside the `app` to do some `async options`. If we pass everything to `app`, we may have cases that we need to duplicate the functions and have to `axios`. One from `pipedreamhq` and the regular NPM package. Please let me know if you disagree

File: components/reddit/reddit.app.mjs
Comment (vellames-turing) [reply components/reddit/reddit.app.mjs:42]: Humm, this is an old bug that you found. Good catch! I will fix it

Comment (vellames-turing) [reply components/reddit/reddit.app.mjs:42]: Fixed!

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (js07) [inline components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs]: It seems like the `limit` parameter of [GET [/r/subreddit]/comments/article](https://www.reddit.com/dev/api/#GET_comments_{article}) correctly limits the number of comments fetched, so we could keep it.

I think most users won't care to set a `limit` if comments are not paginated, so I think removing the prop is fine too.

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (js07) [inline components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs:75]: Remove `after`, `before`, and `limit` from destructure of `this` since the corresponding props have been removed?

```suggestion
```

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (js07) [inline components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs:90]: Remove `after`, `before`, and `limit` params since the corresponding props have been removed?

```suggestion
```

File: components/reddit/reddit.app.mjs
Comment (js07) [inline components/reddit/reddit.app.mjs:118]: **-Minor-**
I think the `options.length > 0` check isn't necessary here because [`_.get`](https://lodash.com/docs/4.17.15#get) will return `undefined` if `${options.length - 1}` is `-1`.

```suggestion
            after: get(options, `${options.length - 1}.value.name`),
```

**Review by js07 — COMMENTED** [review]
The changes and fix to pagination of `async options` for the `subreddit` prop look good. Thanks! I just left a few additional minor comments.

File: components/reddit/actions/list-subreddits-by-query/list-subreddits-by-query.mjs
Comment (js07) [reply components/reddit/actions/list-subreddits-by-query/list-subreddits-by-query.mjs:17]:
```suggestion
    subreddit: {
      propDefinition: [
        reddit,
        "subreddit",
      ],
      description: "Search subreddit titles, and select a subreddit if \"Structured Mode\" is enabled."
```

**Comment by vellames-turing** [comment]
@js07 Ready for review again

File: components/reddit/reddit.app.mjs
Comment (js07) [inline components/reddit/reddit.app.mjs:110]: I think the name of the "display name" property of a subreddit is in snake case.
```suggestion
            value: subreddit.data.display_name,
```

**Review by js07 — COMMENTED** [review]
Looks good to me! Just commented on a small issue.

**Review by js07 — DISMISSED** [review]
Looks good to me

File: components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs
Comment (js07) [inline components/reddit/actions/list-comments-in-a-post/list-comments-in-a-post.mjs:88]: **-Minor-**
There's an extra quote at the end of this summary. I missed it previously.
```suggestion
    $.export("$summary", `Comments for "${get(subredditPost, "label", subredditPost)}" in "${get(subreddit, "label", subreddit)}" subreddit successfully fetched`);
```

File: components/reddit/sources/new-hot-posts-on-a-subreddit/new-hot-posts-on-a-subreddit.mjs
Comment (js07) [inline components/reddit/sources/new-hot-posts-on-a-subreddit/new-hot-posts-on-a-subreddit.mjs:68]: I think this should be `10` rather than a `limit` variable because this source emits a "new event each time a new hot post is added to the top **10** items in a subreddit".

**Review by js07 — COMMENTED** [review]
The changes look good and all of the reddit actions and sources work for me. I just made a couple minor comments.

**Comment by vellames-turing** [comment]
@js07 Suggestions applied. Tks!

**Review by js07 — COMMENTED** [review]
I re-reviewed this PR since the previous approving review was dismissed. New commits after an approving review now cause the review to be dismissed.

I tested the actions and they work for me.

I just left some suggestions for removing references to "Structured Mode" in prop descriptions.

**Comment by dannyroosevelt** [comment]
> I re-reviewed this PR since the previous approving review was dismissed. New commits after an approving review now cause the review to be dismissed.
> 
> I tested the actions and they work for me.
> 
> I just left some suggestions for removing references to "Structured Mode" in prop descriptions.

+1 on removing references to structured mode. And thank you for re-reviewing and testing! Once the suggested changes are implemented, I'm good to publish!

**Comment by vellames-turing** [comment]
@js07 Implemented. Waiting for approval
Cc @dannyroosevelt

**Review by js07 — DISMISSED** [review]
Looks good to me.

**Review by dannyroosevelt — APPROVED** [review]

**Comment by vellames-turing** [comment]
@dannyroosevelt Just checking, this is good to merge?

**Comment by dannyroosevelt** [comment]
> @dannyroosevelt Just checking, this is good to merge?

Yep

**Pull request merged**
